### ISSUE TITLE
Handling of case splits

### ISSUE THREAD

**Issue by martin-neuhaeusser** [mstksg/backprop#13]
While trying to use backprop with ReLU activation functions, I ran into wrong results.
Consider this (admittedly strange) implementation of the identity:

```haskell
f :: (Reifies s W) => BVar s Double -> BVar s Double
f x = if x < 0 then x else x
f' = BP.gradBP f
```

It returns a derivative of 0 both for positive and negative inputs:

```haskell
main :: IO ()
main = do
  putStrLn ("Derivative at -5 = " ++ show (f' (-5)))
  putStrLn ("Derivative at 5 = " ++ show (f' (5)))
```

Executing the code yields
```
Derivative at -5 = 0.0
Derivative at 5 = 0.0
```

Other examples (my original ReLU) produce wrong results, too:
```haskell
f :: (Reifies s W) => BVar s Double -> BVar s Double
f x = if x < 0 then 0 else x
f' = BP.gradBP f
```

Any hints are highly appreciated. Are there any specifics we have to keep in mind when
using case splits in the functions to be differentiated?

**Comment by mstksg** [comment]
Ah yes, this definitely isn't expected behavior.  I'll check into this and let you know what I find, it's likely that it's a regression bug, since I remember explicitly testing something like this in the past.  Thanks for the heads up!

**Comment by mstksg** [comment]
Some further investigation seems to reveal that the problem might be unrelated to case splits.  Consider for instance:

```haskell
f :: (Reifies s W) => BVar s Double -> BVar s Double
f x = 1 * if x < 0 then x else x
f' = BP.gradBP f
```

This returns the correct answers, `f' 100 = 1.0`, and `f' (-100) = 1.0`.

I can isolate the bug further by just passing in `id`

```haskell
BP.gradBP id 100
-- 0
BP.gradBP (* 1) 100
-- 1
```

The bug seems to come from the fact that x is never "touched", which is quite unfortunate :)  I'll look into seeing how this can be fixed, but in the mean time you should be able to get things working in the current version by adding a multiplication by 1 or something to disturb the input.

**Comment by mstksg** [comment]
Should be fixed in 172c2d2! :)  Thanks again for the catch!

EDIT: fix is pushed to hackage as 0.2.6.2

**Comment by martin-neuhaeusser** [comment]
Thank you very much! It works as expected.
